### PR TITLE
Move more RMF code from iOS to BSK

### PR DESCRIPTION
**Pull request by ayoy**
Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL: https://app.asana.com/0/1201621708115095/1207573461751430/f
iOS PR: https://github.com/duckduckgo/iOS/pull/2967
macOS PR: https://github.com/duckduckgo/macos-browser/pull/2878
What kind of version bump will this require?: Major

**Optional**:

CC: @amddg44 

**Description**:
This change moves reusable parts of the Remote Messaging Framework code over from the iOS repository.
The moved parts are: storage class, request class, Core Data model and storage unit tests.

**Steps to test this PR**:
See [iOS PR](https://github.com/duckduckgo/iOS/pull/2967) for steps to test.

**OS Testing**:

* [ ] iOS 14
* [ ] iOS 15
* [ ] iOS 16
* [ ] macOS 10.15
* [ ] macOS 11
* [ ] macOS 12

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: Sources/RemoteMessaging/CoreData/RemoteMessageManagedObject.swift [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+@objc(RemoteMessageManagedObject)
+public class RemoteMessageManagedObject: NSManagedObject {
+
+    @nonobjc public class func fetchRequest() -> NSFetchRequest<RemoteMessageManagedObject> {
+        return NSFetchRequest<RemoteMessageManagedObject>(entityName: "RemoteMessageManagedObject")
+    }
+
+    public class func entity(in context: NSManagedObjectContext) -> NSEntityDescription {
+        return NSEntityDescription.entity(forEntityName: "RemoteMessageManagedObject", in: context)!
+    }
```
Comment (ayoy) [inline Sources/RemoteMessaging/CoreData/RemoteMessageManagedObject.swift:31]: This (and the convenience initializer below) is a boilerplate code also seen in other NSManagedObject subclasses in BSK (e.g. BookmarkEntity).

File: Sources/RemoteMessaging/RemoteMessagingStore.swift [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+import Persistence
+
+public enum RemoteMessagingStoreError: Error {
+    case saveConfigFailed
+    case invalidateConfigFailed
+    case updateMessageShownFailed
+    case saveMessageFailed
+    case updateMessageStatusFailed
+    case deleteScheduledMessageFailed
+}
```
Comment (ayoy) [inline Sources/RemoteMessaging/RemoteMessagingStore.swift:32]: `Pixel.fire` calls have been replaced with `EventMapping` mechanism (also commonly used in BSK), and now whenever a pixel was fired, it's replaced with the event mapping emitting a respective event with an associated error parameter. Client app provides EventMapping class that receives the events and fires pixels.

File: Sources/RemoteMessaging/RemoteMessageRequest.swift [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+//
+
+import Foundation
+import Networking
+
+public struct RemoteMessageRequest {
+
+    public let endpoint: URL
+
+    public init(endpoint: URL) {
```
Comment (ayoy) [inline Sources/RemoteMessaging/RemoteMessageRequest.swift:26]: This struct used to hardcode the endpoint. Now, in order to prepare for reuse by iOS and macOS, it's parametrized with the endpoint URL, provided by the client app.

File: Sources/RemoteMessaging/RemoteMessagingStore.swift [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+    }
+
+    let context: NSManagedObjectContext
+    let notificationCenter: NotificationCenter
+
+    public convenience init(
+        database: CoreDataDatabase,
+        notificationCenter: NotificationCenter = .default,
+        errorEvents: EventMapping<RemoteMessagingStoreError>?,
+        log: @escaping @autoclosure () -> OSLog = .disabled
```
Comment (ayoy) [inline Sources/RemoteMessaging/RemoteMessagingStore.swift:57]: Logger can be provided by the client app.

File: Sources/RemoteMessaging/RemoteMessagingStore.swift [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+
+    private let getLog: () -> OSLog
+    private var log: OSLog {
+        getLog()
+    }
+}
+
+// MARK: - RemoteMessagingConfigManagedObject Public Interface
+
+extension RemoteMessagingStore {
```
Comment (ayoy) [inline Sources/RemoteMessaging/RemoteMessagingStore.swift:106]: All the code below this line has been directly copied over from iOS, with minor adjustments, such as using convenience `fetchRequest()` methods and initializers of NSManagedObject subclasses, as well as using RemoteMessagingConfig convenience initializer with the managed object.

File: Tests/RemoteMessagingTests/RemoteMessagingStoreTests.swift
Comment (ayoy) [inline Tests/RemoteMessagingTests/RemoteMessagingStoreTests.swift:1]: Tests have been moved 1:1 from iOS, with no additions for now.

**Review by amddg44 — COMMENTED** [review]
@ayoy from my perspective all these changes are looking great, nice work!

**Review by dus7 — APPROVED** [review]

**Pull request merged**